scaffold-eth/se-2-challenges
# Challenge 2 extension

I moved challenge 2 since I wanted to try to work with my checklist. Found some mistakes and updated it.

To test: 
```
yarn cli -e scaffold-eth/se-2-challenges:ch2-extension-init
```

Notes:
- despite the fact that we have comment `//Do not change the solidity version as it negativly impacts submission grading` in our contracts, I needed to change version to `0.8.20`. Reason is `openzeppelin` uses [^0.8.20 in Ownable.sol](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/access/Ownable.sol#L4) , and I need to use that `Ownable.sol`. Same issue with most of contracts of `openzeppelin`. Tests works as expected
 
Also changed the text in readme because of that from
```
> 📝 Edit `Vendor.sol` to inherit _Ownable_.

`contract Vendor is Ownable {`

⚠️ You will also need to uncomment the import of Ownable.sol contract!
```
to
```
⚠️ Uncomment the import of Ownable.sol contract!

> 📝 Edit \`Vendor.sol\` to inherit _Ownable_.

\`contract Vendor is Ownable {\`

> 📝 Change constructor of \`Vendor.sol\` to:

\`constructor(address tokenAddress) Ownable(msg.sender) {\`
```




----

Checklist
**1. Branches**
Create two branches in this repo. Target branch should be named like `challenge-[n]-[challenge-name]--extension`, for example `challenge-2-token-vendor--extension`. The working branch name doesn't matter a lot, name it whatever you want.

- [x] Completed

**2. README.md**
Surface structure of extension:
```
├── README.md
└── extension
    ├── README.md.args.mjs
    └── packages
```

Copy `README.md` from current challenge version and place it in the root of extension. This README will be used by the SRE website.

Change this part:
```
git clone https://github.com/scaffold-eth/se-2-challenges.git challenge-0-simple-nft
cd challenge-0-simple-nft
git checkout challenge-0-simple-nft
yarn install
```
to:
```
npx create-eth@latest -e sre-challenge-0 challenge-0-simple-nft
cd challenge-0-simple-nft
```
- [x] Completed

**3. README.md.args.mjs**
Create `README.md.args.mjs` inside `extension` folder. Structure of the file:
```
export const skipQuickStart = true;

export const extraContents = `... `;
```
We're setting `skipQuickStart` to `true` since the default quick start is not very good for us, as in the case of extensions it will be written before the description.

`extraContents` is mostly the same as main `README.md`, except `## Checkpoint 0: 📦 Environment 📚`: node/yarn/git and npx create-eth parts are removed since we don't need those parts from inside an already created instance. Also `yarn chain/deploy/start` is a bit different. 
Use `\` before backticks

- [x] Completed

**4. packages/hardhat**
Add your
- [x] deployments files
- [x] (optional) (?) `openzeppelin` contracts mostly uses `^0.8.20`, so probably you need to change solidity versions in contracts. For example I needed to change version from 0.8.4 to 0.8.20 in challenge 2 since it uses `Ownable` of `openzeppelin`
- [x] (optional) add deps to package.json if needed

**5. packages/nextjs**
Add your
- [x] pages/components/api routes
- [x] `public` folder with `hero.png` and `thumbnail-challenge-n.png` file. Change `n` to number of your challenge. Take these files from `public` folder of that challenge branch in`se-2-challenges` repository. 
Template files are mostly the same. What needed to be changed:
- [x] `/app/layout.tsx.args.mjs`: change number of the challenge in title
- [x] `page.tsx.args.mjs` challenge number inside `externalExtensionName`. Change `description`, take it from return of homepage component of `se-2-challenges` challenge branch
- [x] `/components/Header.tsx.args.mjs` change `menuIconImports`, `menuObjects` and `logoSubtitle`. Don't include home and debug pages
- [x] `/utils/scaffold-eth/getMetadata.ts.args.mjs` change challenge number inside `thumbnailPath`
- [x] `next.config.args.mjs` most likely you don't need that file
- [x] just copy all other `args.mjs` files from challenge 0 PR
- [x] (optional) `package.json` add deps if needed

**6. New template files**
- [ ] create new branch on `create-eth` with your `template.mjs` files
- [ ] branch from your second branch (see point 1), so it will be third branch of your challenge. Add `.args.mjs` files
- [ ] merge both branches when approved

If I forgot something or if you find something new please add it to this checklist

Thank you!
